packages/@aws-cdk/aws-synthetics/lib/canary.ts

```diff
@@ -0,0 +1,255 @@
+          'cloudwatch:PutMetricData',
+          'logs:CreateLogGroup',
+          'logs:CreateLogStream',
+          'logs:PutLogEvents',
+        ],
+      })],
+    });
+    const inlinePolicies = { canaryPolicy: policy };
+
+    this.role = props.role ?? new iam.Role(this, 'ServiceRole', {
+      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+      inlinePolicies,
+    });
+
+    const duration = props.timeToLive ?? cdk.Duration.seconds(0);
+    const expression = props.frequency ?? cdk.Duration.minutes(5);
+    var timeout = props.timeout ?? cdk.Duration.seconds(Math.min(expression.toSeconds(), 900));
+    timeout = cdk.Duration.seconds(Math.min(timeout.toSeconds(), expression.toSeconds()));
+
+    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
+      artifactS3Location: s3Location,
+      executionRoleArn: this.role.roleArn,
+      startCanaryAfterCreation: props.enableCanary ?? true,
+      runtimeVersion: 'syn-1.0',
+      name: this.verifyName(props.name),
+      runConfig: {
+        // Will include MemorySize when generated code gets updated.
+        timeoutInSeconds: timeout.toSeconds(),
+      },
+      schedule: {
+        durationInSeconds: String(duration.toSeconds()),
+        expression: this.createExpression(expression),
+      },
+      failureRetentionPeriod: props.failureRetentionPeriod?.toDays(),
+      successRetentionPeriod: props.successRetentionPeriod?.toDays(),
+      code: {
+        handler: this.verifyHandler('index.handler'),
+        script: 'foo',
+      },
+    });
+    resource.node.addDependency(this.role);
+
+    this.canaryId = resource.attrId;
+    this.canaryState = resource.attrState;
+    this.canaryName = this.getResourceNameAttribute(resource.ref);
+  }
+
+  /**
+   * Creates the expression that is readable by CloudFormation
+   *
+   * @param frequency - the provided frequency
+   */
+  private createExpression(frequency: cdk.Duration): string {
+    this.verifyExpression(frequency);
+    if (frequency.toMinutes() === 1) {
+      return 'rate(1 minute)';
+    }
+    return `rate(${frequency.toMinutes()} minutes)`;
+  }
+
+  /**
+   * Verifies that the given expression evaluates to 0 or is inside the range of 1-60 minutes.
+   *
+   * @param frequency - the provided frequency
+   */
+  private verifyExpression(frequency: cdk.Duration) {
+    if (frequency.toMinutes() !== 0 && (frequency.toMinutes() < 1 || frequency.toMinutes() > 60)) {
+      throw new Error('Frequency must be between 1 minute and 1 hour');
+    }
+  }
+
+  /**
+   * Verifies that the given handler ends in '.handler'. Returns the handler if successful and
+   * throws an error if not.
+   *
+   * @param handler - the handler given by the user
+   */
+  private verifyHandler(handler: string): string {
+    if (handler.split('.').length !== 2 || handler.split('.')[1] !== 'handler') {
+      throw new Error('Canary Handler must end in \'.handler\'');
+    }
+    if (handler.length > 21) {
```

Comment:
Where does this restriction come from? The [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-synthetics-canary-code.html#cfn-synthetics-canary-code-handler) say the max is 1024.

Reply:
The docs are incorrect (afaik). I've verified that creating a canary with a long name (>21) results in the error "invalid source handler". Leaving this as unresolved for now because perhaps we should update the docs, but this error was the result of a painful investigation on my part.